boa_engine/src/builtins/intl/mod.rs

```diff
@@ -144,3 +147,566 @@ impl Intl {
+        if !has_key {
+            keywords.push(keyword_val);
+        }
+    };
+
+    // 8. Return the Record { [[Attributes]]: attributes, [[Keywords]]: keywords }.
+    UniExtRecord {
+        attributes,
+        keywords,
+    }
+}
+
+fn js_arr_contains_str(js_arr: &JsValue, str_to_find: &str, context: &mut Context) -> bool {
+    let js_arr_obj = js_arr
+        .to_object(context)
+        .unwrap_or_else(|_| JsObject::empty());
+    let js_arr_len = js_arr_obj.length_of_array_like(context).unwrap_or(0);
+    for index in 0..js_arr_len as u32 {
+        let arr_item = js_arr_obj
+            .get(index, context)
+            .unwrap_or_else(|_| JsValue::undefined())
+            .to_string(context)
+            .unwrap_or_else(|_| JsString::new(""))
+            .to_string();
+
+        if arr_item.eq(str_to_find) {
+            return true;
+        }
+    }
+
+    false
+}
+
+/// The `InsertUnicodeExtensionAndCanonicalize` abstract operation inserts `extension`, which must
+/// be a Unicode locale extension sequence, into `locale`, which must be a String value with a
+/// structurally valid and canonicalized Unicode BCP 47 locale identifier.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma402/#sec-insert-unicode-extension-and-canonicalize
+fn insert_unicode_extension_and_canonicalize(locale: &str, extension: &str) -> JsString {
+    // TODO 1. Assert: locale does not contain a substring that is a Unicode locale extension sequence.
+    // TODO 2. Assert: extension is a Unicode locale extension sequence.
+    // TODO 3. Assert: tag matches the unicode_locale_id production.
+    // 4. Let privateIndex be ! StringIndexOf(locale, "-x-", 0).
+    let private_index = locale.find("-x-");
+    let new_locale = match private_index {
+        // 5. If privateIndex = -1, then
+        None => {
+            // a. Let locale be the string-concatenation of locale and extension.
+            locale.to_owned() + extension
+        }
+        // 6. Else,
+        Some(idx) => {
+            // a. Let preExtension be the substring of locale from position 0, inclusive,
+            // to position privateIndex, exclusive.
+            let pre_extension = &locale[0..idx];
+
+            // b. Let postExtension be the substring of locale from position privateIndex to
+            // the end of the string.
+            let post_extension = &locale[idx..];
+
+            // c. Let locale be the string-concatenation of preExtension, extension,
+            // and postExtension.
+            pre_extension.to_owned() + extension + post_extension
+        }
+    };
+
+    // TODO 7. Assert: ! IsStructurallyValidLanguageTag(locale) is true.
+    // 8. Return ! CanonicalizeUnicodeLocaleId(locale).
+    Intl::canonicalize_locale(&new_locale)
+}
+
+#[derive(Debug)]
+pub struct ResolveLocaleRecord {
+    pub locale: JsString,
+    pub properties: HashMap<JsString, JsValue>,
```

Comment:
I think all of the `properties` keys in these three structs represent `[[LocaleData]]` slots from the spec right? I think they can be more strict to be easier to work with. The spec defines the slots here: https://tc39.es/ecma402/#sec-internal-slots.
I think `[[LocaleData]]` could probably be `FxHashMap<JsString, FxHashMap<JsString, Vec<JsString>>>`. That should also make the `js_arr_contains_str` function obsolete and remove all of the JsValue and JsObject usage.

Reply:
https://github.com/boa-dev/boa/pull/2036/files#diff-1dd4af22d3a6142a8060d056e9d0489d84d8a4d9c68a8f6e7b359b8d3ab35499R584

Reply:
Yeah true that seems to be a value according to spec. Let's leave it at this and maybe we can figure out optimizations when we have more implemented and directly see how the code works together.